### PR TITLE
drivers: sensor: bh1750: Add driver for BH1750

### PR DESCRIPTION
**Pull request by heitzmann**
Add driver for the Rohm BH1750 16 bit light sensor.  The sensor operates
in one of 3 resolution modes, selected thru Kconfig, and the driver
allows runtime configuration of the oversampling factor.

Signed-off-by: Lucas Heitzmann Gabrielli <heitzmann@gmail.com>

### CODE REVIEW

File: drivers/sensor/bh1750/Kconfig [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+choice
+	prompt "BH1750 resolution mode"
+	default BH1750_NORMAL_RES
+	help
+	  Select BH1750 resolution mode. The illuminance per LSB in normal and
+	  low resolution modes is mode is 57.5 / oversampling-factor (in lux),
+	  with typical sensitivities of 1 lx for normal and 4 lx for low res.
+	  In high res., the illuminance per LSB is 28.75 / oversampling-factor
+	  and the typical sensitivity is 0.5 lx.
+config BH1750_NORMAL_RES
```
Comment (ycsin) [inline drivers/sensor/bh1750/Kconfig:21]: one empty line above this line

File: drivers/sensor/bh1750/Kconfig [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	bool "normal resolution"
+config BH1750_HIGH_RES
```
Comment (ycsin) [inline drivers/sensor/bh1750/Kconfig:23]: one empty line above this line

File: drivers/sensor/bh1750/Kconfig [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	bool "high resolution"
+config BH1750_LOW_RES
```
Comment (ycsin) [inline drivers/sensor/bh1750/Kconfig:25]: one empty line above this line

File: drivers/sensor/bh1750/bh1750.c [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+static int bh1750_attr_set(const struct device *dev, enum sensor_channel chan,
+			   enum sensor_attribute attr, const struct sensor_value *val)
+{
+	int ret;
+	uint8_t byte;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_LIGHT || chan == SENSOR_CHAN_ALL);
+	__ASSERT_NO_MSG(attr == SENSOR_ATTR_OVERSAMPLING);
+
+	if (val->val1 < 31 || val->val1 > 254) {
```
Comment (ycsin) [inline drivers/sensor/bh1750/bh1750.c:39]: use a macro instead of magic number

File: drivers/sensor/bh1750/bh1750.c [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+		LOG_ERR("Oversampling factor out of range [31, 254]: %d", val->val1);
+		return -EINVAL;
+	}
+
+	byte = BH1750_MEAS_TIME_MSB | (((uint8_t)val->val1 >> 5) & 0x07);
+	ret = bh1750_write(dev, byte);
+	if (ret < 0) {
+		LOG_ERR("Failed to configure high part of oversampling factor.");
```
Comment (ycsin) [inline drivers/sensor/bh1750/bh1750.c:47]:
```suggestion
		LOG_ERR("Failed to configure %s part of oversampling factor.", "high");
```

You can save a few bytes if you do it like so, same goes for the low part below

File: drivers/sensor/bh1750/bh1750.c [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+	k_busy_wait(BH1750_TIME_FACTOR * (uint32_t)data->oversampling_factor);
+
+	uint8_t buffer[2];
+	ret = bh1750_read(dev, buffer, 2);
+	if (ret < 0) {
+		LOG_ERR("Unable to issue read command.");
+		return ret;
+	}
+	data->adc_count = ((uint16_t)buffer[0] << 8) | (uint16_t)buffer[1];
```
Comment (ycsin) [inline drivers/sensor/bh1750/bh1750.c:97]: theres a function for this
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/byteorder.h#L377

File: drivers/sensor/bh1750/bh1750.c [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+	return 0;
+}
+
+static int bh1750_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_LIGHT);
+	struct bh1750_data *data = dev->data;
+	uint32_t tmp = (uint32_t)data->adc_count * (230 << 8) / data->oversampling_factor;
```
Comment (ycsin) [inline drivers/sensor/bh1750/bh1750.c:107]: one empty line after var declaration

File: drivers/sensor/bh1750/bh1750.c [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+#if defined(CONFIG_BH1750_HIGH_RES)
+	/* tmp = lux * 2^11 */
+	val->val1 = (int32_t)(tmp >> 11);
+	val->val2 = (int32_t)((tmp & 0x07FF) * 15625) >> 5;
```
Comment (ycsin) [inline drivers/sensor/bh1750/bh1750.c:111]: use macro to better document these magic numbers?

File: drivers/sensor/bh1750/bh1750.h [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+#define BH1750_RESET 0x07
+
+#define BH1750_CMD_HRES 0x20
+#define BH1750_CMD_HRES2 0x21
+#define BH1750_CMD_LRES 0x23
+
+#if defined(CONFIG_BH1750_NORMAL_RES)
+#define BH1750_SINGLE_MEASUREMENT BH1750_CMD_HRES
+#define BH1750_TIME_FACTOR ((uint32_t)2610)
```
Comment (ycsin) [inline drivers/sensor/bh1750/bh1750.h:23]: I think you only need `2610U`?

File: dts/bindings/sensor/rohm,bh1750-i2c.yaml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    type: int
+    required: false
+    description: |
+      Adjust the oversampling factor of the sensor to compensate for window
+      absorption. For example, if the sensor is behind an optical window with
+      50% absorption, the default factor should be multiplied by 2 (138).
+      Acceptable values are from 31 to 254, corresponding to (0.45 and 3.68
+      times the default value, respectively).  The sampling time is this factor
+      multiplied by 2.61 ms (normal and high resolution) or 0.35 ms (low res).
+    default: 69
```
Comment (ycsin) [inline dts/bindings/sensor/rohm,bh1750-i2c.yaml:21]: You might have to reason why 69 is the default

**Review by ycsin — CHANGES_REQUESTED** [review]

File: drivers/sensor/bh1750/Kconfig [hunk 11]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2022 Lucas Heitzmann Gabrielli
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig BH1750
+	bool "Rohm BH1750 16 bit ambient light sensor"
+	depends on I2C
```
Comment (galak) [inline drivers/sensor/bh1750/Kconfig:6]: add:
```
default y
depends on DT_HAS_ROHM_BH1750_ENABLED
```

File: dts/bindings/sensor/rohm,bh1750-i2c.yaml [hunk 12]
```diff
@@ -0,0 +1,21 @@
+# Copyright (c) 2022 Lucas Heitzmann Gabrielli
```
Comment (galak) [inline dts/bindings/sensor/rohm,bh1750-i2c.yaml:1]: rename file to just `rohm,bh1750.yaml`.  We use the `-i2c` when a device supports different buses like `i2c` and  `spi`.

**Review by galak — CHANGES_REQUESTED** [review]
* Kconfig update
* rename binding yaml file.

File: drivers/sensor/bh1750/bh1750.c (hunk 8 above)
Comment (heitzmann) [reply drivers/sensor/bh1750/bh1750.c:111]: The numbers here come from a mix of the datasheet equations and shifts to preserve the most significant digits without increasing computational cost (or risking overflow). I couldn't think of macro names that would properly document them, so I added actual documentation to the function, if that's okay.

File: dts/bindings/sensor/rohm,bh1750-i2c.yaml (hunk 10 above)
Comment (heitzmann) [reply dts/bindings/sensor/rohm,bh1750-i2c.yaml:21]: I've improved the description in more accurate terms and included the reason for the default value (it is the reset default).

**Comment by heitzmann** [comment]
Thanks @galak and @ycsin for the reviews and comments! I believe I've implemented all the changes.

File: drivers/sensor/bh1750/bh1750.h [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+#endif
+
+#define BH1750_MEAS_TIME_MSB 0x40
+#define BH1750_MEAS_TIME_LSB 0x60
+
+#define BH1750_MEAS_TIME_MIN 31
+#define BH1750_MEAS_TIME_MAX 254
+
+struct bh1750_config {
+	const struct device *bus;
```
Comment (galak) [inline drivers/sensor/bh1750/bh1750.h:39]: Make this use `struct i2c_dt_spec i2c;` instead.

See `drivers/sensor/adt7420` for an example

File: drivers/sensor/bh1750/bh1750.c [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+		LOG_ERR("Bus device is not ready");
+		return -ENODEV;
+	}
+
+	return bh1750_attr_set(dev, SENSOR_CHAN_ALL, SENSOR_ATTR_OVERSAMPLING, &tmp);
+}
+
+#define BH1750_INST(inst)                                                                          \
+	static const struct bh1750_config bh1750_config_##inst = {                                 \
+		.bus = DEVICE_DT_GET(DT_INST_BUS(inst)),                                           \
```
Comment (galak) [inline drivers/sensor/bh1750/bh1750.c:174]: replace with `I2C_DT_SPEC_INST_GET()` and `struct i2c_dt_spec` usage.

File: drivers/sensor/bh1750/bh1750.c [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+
+LOG_MODULE_REGISTER(BH1750, CONFIG_SENSOR_LOG_LEVEL);
+
+static int bh1750_read(const struct device *dev, uint8_t *byte, uint32_t count)
+{
+	const struct bh1750_config *cfg = dev->config;
+
+	return i2c_read(cfg->bus, byte, count, cfg->bus_addr);
```
Comment (galak) [inline drivers/sensor/bh1750/bh1750.c:23]: use i2c_dt_spec and i2c_read_dt

File: drivers/sensor/bh1750/bh1750.c [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+static int bh1750_write(const struct device *dev, uint8_t byte)
+{
+	const struct bh1750_config *cfg = dev->config;
+
+	return i2c_write(cfg->bus, &byte, 1, cfg->bus_addr);
```
Comment (galak) [inline drivers/sensor/bh1750/bh1750.c:30]: use i2c_dt_spec and i2c_write_dt

**Review by galak — CHANGES_REQUESTED** [review]
Update to use i2c_dt_spec and associated APIs

File: drivers/sensor/bh1750/bh1750.c (hunk 8 above)
Comment (ycsin) [reply drivers/sensor/bh1750/bh1750.c:143]: Is there a name for the bit/register position/mask?

Comment (heitzmann) [reply drivers/sensor/bh1750/bh1750.c:143]: No, the datasheet doesn't describe the registers as usual. It only mentions the data register (that holds the last ADC count, which is the name I used) and the measurement time register (which I renamed to oversampling factor to keep in line with the zephyr convention). The IC is very simple; I don't believe there are any other registers.

**Comment by morsisko** [comment]
I see that we kinda did the same job twice: https://github.com/zephyrproject-rtos/zephyr/pull/43086

File: drivers/sensor/bh1750/Kconfig (hunk 11 above)
Comment (MaureenHelm) [reply drivers/sensor/bh1750/Kconfig:8]: This still needs to be done

File: drivers/sensor/bh1750/Kconfig [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+	default y
+	depends on DT_HAS_ROHM_BH1750_ENABLED
+	depends on I2C
+	help
+	  Enable driver for the Rohm BH1750 pressure/altitude sensor
+
+if BH1750
+
+choice
+	prompt "BH1750 resolution mode"
```
Comment (MaureenHelm) [inline drivers/sensor/bh1750/Kconfig:15]: This option should be converted to a devicetree property to allow configuring it on a per-instance basis

File: drivers/sensor/bh1750/bh1750.c [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+	struct bh1750_data *data = dev->data;
+
+	val->val1 = (int32_t)data->oversampling_factor;
+
+	return 0;
+}
+
+static int bh1750_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_LIGHT);
```
Comment (MaureenHelm) [inline drivers/sensor/bh1750/bh1750.c:85]: Return an errno

File: drivers/sensor/bh1750/bh1750.c [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+	}
+	data->adc_count = sys_get_be16(buffer);
+
+	return 0;
+}
+
+static int bh1750_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_LIGHT);
```
Comment (MaureenHelm) [inline drivers/sensor/bh1750/bh1750.c:112]: Return an errno

File: dts/bindings/sensor/rohm,bh1750.yaml [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+      Adjust the sensor sensitivity. The sensitivity can be calculated as
+      57.5 / oversampling-factor (in lux per LSB) for normal and low resolution
+      modes and 28.75 / oversampling-factor for high resolution. The default
+      value on reset is 69, which gives 0.83 lx per LSB on normal and low
+      resolution modes and 0.42 lx per LSB on high resolution. Higher values
+      improve sensitivity but increase the sampling interval, calculated as
+      2.61 ms * oversampling-factor for high and normal resolution modes, and
+      as 0.35 ms * oversampling-factor for low resolution. The minimal and
+      maximal values are 31 and 254.
+    default: 69
```
Comment (MaureenHelm) [inline dts/bindings/sensor/rohm,bh1750.yaml:24]: Why was this default value chosen? Is it a power-on register reset value?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please squash the fixup commits into the original commit

**Pull request closed without merging**